src/hotspot/share/utilities/rbTree.hpp

```diff
@@ -158,95 +168,126 @@ class RBTree {
   // Removes the node with the given key from the tree if it exists.
   // Returns true if the node was successfully removed, false otherwise.
   bool remove(const K& k) {
-    RBNode* node = find_node(_root, k);
-    return remove(node);
+    RBNode* node = find_node(k);
+    if (node == nullptr){
+      return false;
+    }
+    remove(node);
+    return true;
   }
 
-  // Removes the given node from the tree.
-  // Returns true if the node was successfully removed, false otherwise.
-  bool remove(RBNode* node);
+  // Removes the given node from the tree. node must be a valid node
+  void remove(RBNode* node);
 
   // Removes all existing nodes from the tree.
   void remove_all() {
-    GrowableArrayCHeap<RBNode*, mtInternal> to_delete(2 * log2i(_num_nodes + 1));
-    to_delete.push(_root);
+    RBNode* to_delete[64];
+    int stack_idx = 0;
+    to_delete[stack_idx++] = _root;
 
-    while (!to_delete.is_empty()) {
-      RBNode* head = to_delete.pop();
+    while (stack_idx > 0) {
+      RBNode* head = to_delete[--stack_idx];
       if (head == nullptr) continue;
-      to_delete.push(head->_left);
-      to_delete.push(head->_right);
+      to_delete[stack_idx++] = head->_left;
+      to_delete[stack_idx++] = head->_right;
       free_node(head);
     }
     _num_nodes = 0;
     _root = nullptr;
   }
 
+  // Alters behaviour of closest_(leq/gt) functions to include/exclude the exact value
+  enum BoundMode : uint8_t { EXCLUSIVE, INCLUSIVE };
+
   // Finds the node with the closest key <= the given key
-  RBNode* closest_leq(const K& key) {
+  // Change mode to EXCLUSIVE to not include node matching key
+  RBNode* closest_leq(const K& key, BoundMode mode = INCLUSIVE) {
```

Comment:
Having a parameter value change the behaviour to contradict the name of the function feels a little evil. Might cause confusion and bugs in the future. I would rather have one `closest(Key, Direction, BoundMode)` and/or four `closest_[ls,leq,geq,gt](Key)`.